### PR TITLE
Criteria implementation

### PR DESCRIPTION
**Pull request by evsign**
Promised pull request)
I was very wondered there is no tests and with this pr fix some of that) I cover all my criteria code and `findAll` method in `EloquentRepository`. In future i want to cover all and hope somebody help with it)

And maybe somebody can write documentation for this, because i afraid my poor english does a bad job.

How to use:
At first add `Rinvex\Repository\Traits\Criteriable` trait in your repository. 
```php
use Rinvex\Repository\Repositories\EloquentRepository;
use Rinvex\Repository\Traits\Criteriable;

class UserRepository extends EloquentRepository
{
    use Criteriable;

    protected $model = User::class;
    protected $repositoryId = 'rinvex.repository.user';
}
```
After that you can `pushCriteria` or `pushCriterion` to your repository. Criteria can be implemented as separeted classes which must implement `Rinvex\Repository\Contracts\CriterionContract` with one method `apply($query, $reposiroty)`. Criterion can be pushed by class name, like instantiated object or closure which must have same signature like `apply` method in `CriterionContract`.
```php
$repository->pushCriteria([new RangeCriterion, new CheckCriterion]);
$repository->pushCriteria([RangeCriterion::class, CheckCriterion::class]);
$repository->pushCriteria([new RangeCriterion, function ($query, $repository) { return $query->where('id', 1); }]);
$repository->pushCriterion(function ($query, $repository) { return $query->where('id', 1); });
```
Applying occurs during execution `prepareQuery` method in `BaseRepository`.

If you want perform query without pushed criteria you can use `$repository->skipCriteria()->findAll()` method. Also you can add  and skip by analogy default criteria. `$repository->setDefaultCriteria([])->skipDefaultCriteria()`

Also there is some helpers methods for checking and getting criteria.
```php
$repository->hasCriterion(RangeCriterion::class); //check for existanse of given criterion in current criteria list
$repository->getCriterion(RangeCriterion::class); //return criterion object from criteria list
$repository->removeCriterion(RangeCriterion::class); //remove criterion of this clas from current criteria list
```

Some notes about closure criteria. If you pass in arguments closure criterion declaration, there is no way to remove or check for this creterion existence. This is normal and describing this behavior is beyond of this message)

Also you can use this trait in anything places where you want to implement criteriable behavior.
For example you can use it in eloquent model, but you must find the way to apply criteria.
```php
use Illuminate\Database\Eloquent\Model;
use Rinvex\Repository\Traits\Criteriable;

class User extends Model
{
    use Criteriable;

    public function scopeCriteriable($query)
    {
        return $this->applyCriteria($query, $this);
    }
}

$user = new User;
$user->pushCriterion(RangeCriterion::class);
$user->criteriable()->get();
```
#60

### CODE REVIEW

**Comment by ionut-tanasa** [comment]
Awesome!

One doubt tho. Did not go through the code yet, but what's the diff between Criteria or Criterion?

**Comment by evsign** [comment]
@ionut-tanasa its just singular `criterion` and plural `criteria` forms.

**Comment by Omranic** [comment]
Oh, wow 😮 
Impressive PR, I'll check it soon -hopefully this week-..
I appreciate adding test suites too, much much appreciated @evsign 😄

File: src/Traits/Criteriable.php [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+     * @return $this
+     */
+    protected function addCriterion($criterion, $list)
+    {
+        if (property_exists($this, $list) &&
+            ($criterion instanceof Closure ||
+                $criterion instanceof CriterionContract ||
+                (is_string($criterion) &&
+                    class_exists($criterion) &&
+                    ($criterion = new $criterion()) instanceof CriterionContract
```
Comment (Solodkiy) [inline src/Traits/Criteriable.php]: Might be fatal error here if class constructor have required arguments.
I suggest use ReflectionClass.

File: src/Traits/Criteriable.php [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+                ))
+        ) {
+            $this->$list[$this->getCriterionName($criterion)] = $criterion;
+        }
+
```
Comment (Solodkiy) [inline src/Traits/Criteriable.php]: No reaction to invalid argument is ok?
Might be better throws exception in this case?

File: src/Traits/Criteriable.php (hunk 1 above)
Comment (evsign) [inline src/Traits/Criteriable.php]: Yes. I think about passing criteria by class name only for simple criteria without constructor, but give possibility to pass corresponding arguments in array could be nice. For example
```php
[
        new TestCriterion(),
        RangeCriterion::class => ['from' => '2016-08-01', 'to' => '2016-08-30']
]
```
Rework this at evening) Thanks.

File: src/Traits/Criteriable.php (hunk 2 above)
Comment (evsign) [inline src/Traits/Criteriable.php]: Sure, thanks. Wanted, but forgot.

File: src/Traits/Criteriable.php (hunk 1 above)
Comment (Solodkiy) [inline src/Traits/Criteriable.php]: IMHO `RangeCriterion::class => ['from' => '2016-08-01', 'to' => '2016-08-30']` is unnecessary magic. `new RangeCriterion('2016-08-01',  '2016-08-30')` is better.

Comment (evsign) [inline src/Traits/Criteriable.php]: I think too, but my part just provide the way, and everybody itself decide what is better)

**Pull request merged**